# Introduction to pandas for Exploratory Data Analysis

## Learning Objectives

In this notebook, you will:

- Understand the typical Data Science workflow, with emphasis on **Exploratory Data Analysis (EDA)**.
- Learn the fundamental concepts of the **pandas** library for data manipulation and analysis.
- Load, inspect, transform, and export datasets in multiple formats.
- Develop initial skills to prepare datasets for exploratory analysis and visualization.

---
## The Role of pandas in the Data Science Workflow

A typical data analysis workflow usually includes:

### 1. ETL (Extract, Transform, Load)
- Extract data from multiple sources.
- Clean and transform raw data.
- Organize data into a suitable analytical format.

### 2. Exploratory Data Analysis (EDA)
- Initial understanding of the dataset.
- Identification of patterns, inconsistencies, and outliers.
- Hypothesis generation.

### 3. Data Visualization
- Graphical representation to support interpretation.

### 4. Communication of Results
- Reports, dashboards, or presentations.

**pandas** is one of the main tools used during ETL and EDA stages.

---
## What is pandas?

**pandas** is a widely used Python library designed for:

- Structured data manipulation.
- Data cleaning and preprocessing.
- Fast exploratory data analysis.
- Integration with visualization and machine learning tools.

### Core Data Structures

- **Series** → One-dimensional labeled array.
- **DataFrame** → Two-dimensional tabular data structure (most commonly used).

---
## Reading and Writing Data with pandas

pandas supports several data formats.

### Common Input Formats

- CSV (`read_csv`)
- Excel (`read_excel`)
- JSON (`read_json`)


### Exporting Data

- `to_csv()`
- `to_excel()`
- `to_json()`
-
Correctly importing data is a critical step in any exploratory analysis.

---
## Typical EDA Tasks Using pandas

Some common operations include:

- Previewing data (`head()`, `tail()`).
- Inspecting structure (`info()`).
- Generating descriptive statistics (`describe()`).
- Handling missing values.
- Filtering and grouping data.
- Creating new features or variables.
- Preparing datasets for visualization.

---

## Jupyter Notebooks

![Jupyter Logo](https://jupyter.org/assets/homepage/main-logo.svg)

Jupyter Lab/Notebooks will be our IDE of preference.

Jupyter Notebooks are composed by Cells.

Cells can be either **Markdown** or **Code**.

- **Markdown** cells allows us to write text, formulas, add images and explain things in the same way we would do in a normal notebook. Here is a [cheat sheet](https://guides.github.com/pdfs/markdown-cheatsheet-online.pdf) for Markdown Sintax.

- **Code** Cells allows us to execute code (usually _Python_).

Here is a Jupyter Notebook commands [cheat sheet](https://www.edureka.co/blog/wp-content/uploads/2018/10/Jupyter_Notebook_CheatSheet_Edureka.pdf) although you can also use the Help tab.

(check also these 28 _advanced_ Jupyter Notebook Tips, Tricks, and Shortcuts: https://www.dataquest.io/blog/jupyter-notebook-tips-tricks-shortcuts/)

Useful Jupyter Notebook shortcuts:
| **Action** | **Jupyter Notebook** | **Google Colab** |
|-|:-:|:--:|
| Add a cell above | A | Ctrl + M + A |
| Add a cell below | B | Ctrl + M + B |
| See all keyboard shortcuts | H | Ctrl + M + H |
| Change cell to code | Y | Ctrl + M + Y |
| Change cell to markdown | M | Ctrl + M + M |
| Interrupt the kernel | II | Ctrl + M + I |
| Delete a cell | DD | Ctrl + M + D |
| Execute cell and go to the next | Shift + Enter | Shift + Enter |
| Execute cell and remain on current | Ctrl + Enter | Ctrl + Enter |


## Rectangular data

A lot of the datasets we study in Data Science are tabular, like a spreadsheet or database table. This means that we have a  `n` and `m` columns.

* A **column** in the table is commonly referred to as a **variable**.
    * You may also see them refered as: attribute, input, predictor or feature
* A **row** in the table is commonly referred to as a **record**.
    * You may also see them refered as: case, example, instance, observation, pattern or sample


<details>
<p>

> Rectangular data is essentially a two-dimensional matrix with rows indicating records and columns indicating variables
> -- <cite>[Peter Bruce, Andrew Bruce @ Practical Statistics for Data Scientists][3]</cite>

[3]: https://www.oreilly.com/library/view/practical-statistics-for/9781491952955/ch01.html

</p>
</details>


# Pandas Basics

What is Pandas?

<details open>
<p>

> Pandas is seriously a game changer when it comes to cleaning, transforming, manipulating and analyzing data. In simple terms, Pandas helps to clean the mess
> -- <cite>[Admond Lee][1]</cite>

[1]: https://towardsdatascience.com/why-and-how-to-use-pandas-with-large-data-9594dda2ea4c

</p>
</details>

Have a look at the Pandas documentation

<details open>
<p>

> `pandas` is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool,
built on top of the Python programming language.
> -- <cite>[pandas library][2]</cite>

[2]: https://pandas.pydata.org/

</p>

- API reference: https://pandas.pydata.org/docs/reference/index.html#api

In [2]:
import pandas as pd

## Pandas Series and DataFrames

### Pandas Series
Data structure that stores data in the form of a column. A series is normally used to store information about a particular attribute in your dataset.

In [3]:
my_list = [10, 20, 30, 40, 50]
# print(my_list[2])

my_series = pd.Series(my_list)
print(my_series)

0    10
1    20
2    30
3    40
4    50
dtype: int64


In the above output, you can see that the indexes for a series start from 0 to 1 less than the number of items in the series.

You can also define custom indexes for your series. To do so, you need to pass your list of indexes to the index attribute of the Series class, as shown in the script below:

In [4]:
my_series = pd.Series([10, 20, 30, 40, 50], index=["num1", "num2", "num3", "num4", "num5"])
print(my_series)

num1    10
num2    20
num3    30
num4    40
num5    50
dtype: int64


Finally, you can also create a series using a dictionary.

In this case, the dictionary keys will become series indexes while the dictionary values are inserted as series items.

Here is an example:

In [5]:
my_dict = {'num1': 6,
           'num2': 7,
           'num3': 8}
my_series = pd.Series(my_dict)
print(my_series)

num1    6
num2    7
num3    8
dtype: int64


#### Useful Operations on Pandas Series

In [6]:
## Accessing Items

my_series = pd.Series([10, 20, 30, 40, 50], index=["num1", "num2", "num3", "num4", "num5"])
print(my_series['num4'])
print(my_series['num3'])

40
30


In [7]:
## Finding Maximum and Minimum Values
print("Maximum value:", my_series.max())
print("Minimum value:", my_series.min())

Maximum value: 50
Minimum value: 10


In [8]:
## Others
print("Sum:", my_series.sum())
print("Mean:", my_series.mean())
print("Standard Deviation:", my_series.std())

## Finding Data Type
print("Data Type:", my_series.dtype)

## Converting to List
print("As List:", my_series.tolist())

Sum: 150
Mean: 30.0
Standard Deviation: 15.811388300841896
Data Type: int64
As List: [10, 20, 30, 40, 50]


### Pandas Dataframe

A Pandas dataframe is a tabular data structure that stores data in the form of rows and columns. As a standard, the rows correspond to records while columns refer to attributes.

In simplest words, a Pandas dataframe is a collection of series.


In [9]:
# dataframe using list of lists

scores = [['Mathematics', 85], ['English', 91], ['History', 95]]
my_df = pd.DataFrame(scores, columns=['Subject', 'Score'])
my_df

,Subject,Score
0,Mathematics,85
1,English,91
2,History,95


In [10]:
# dataframe using dictionaries
scores = {
    'Subject': ["Mathematics", "History", "English", "Science", "Arts"],
    'Score': [98, 75, 68, 82, 99]
}
my_df = pd.DataFrame(scores)
my_df

,Subject,Score
0,Mathematics,98
1,History,75
2,English,68
3,Science,82
4,Arts,99


In [11]:
# dataframe using list of dictionaries
scores = [
    {'Subject': 'Mathematics', 'Score': 85},
    {'Subject': 'History', 'Score': 98},
    {'Subject': 'English', 'Score': 76},
    {'Subject': 'Science', 'Score': 72},
    {'Subject': 'Arts', 'Score': 95},
]
my_df = pd.DataFrame(scores)
my_df

,Subject,Score
0,Mathematics,85
1,History,98
2,English,76
3,Science,72
4,Arts,95


The dictionaries within the list used to create a Pandas dataframe need not be of the same size.

In such cases, the missing values are filled with NaN (Not a Number) by default.

In [12]:
scores = [
    {'Subject': 'Mathematics', 'Score': 85},
    {'Subject': 'History', 'Score': 98},
    {'Subject': 'English', 'Score': 76},
    {'Subject': 'Science'},
    {'Subject': 'Arts', 'Score': 95},
]
my_df = pd.DataFrame(scores)
my_df

,Subject,Score
0,Mathematics,85.0
1,History,98.0
2,English,76.0
3,Science,NaN
4,Arts,95.0


In [13]:
my_df['Score']

0    85.0
1    98.0
2    76.0
3     NaN
4    95.0
Name: Score, dtype: float64

#### Basic operations on DataFrames

In [14]:
## head() and tail()

### display vs print
# display(my_df.head())
#
display(my_df.tail())

,Subject,Score
0,Mathematics,85.0
1,History,98.0
2,English,76.0
3,Science,NaN
4,Arts,95.0


In the output below, you can see the number of entries in your Pandas dataframe, the number of columns along with their column type, and so on.

In [15]:
# getting dataframe info

my_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Subject  5 non-null      str    
 1   Score    4 non-null      float64
dtypes: float64(1), str(1)
memory usage: 212.0 bytes


Finally, to get information such as mean, minimum, maximum, standard deviation, etc., for **numeric** columns in your Pandas dataframe, you can use the describe() method

In [16]:
my_df.describe()

,Score
count,4.000000
mean,88.500000
std,10.016653
min,76.000000
25%,82.750000
50%,90.000000
75%,95.750000
max,98.000000


## Importing Data in Pandas

- `pd.read_csv()`: reading delimited files.
- `pd.read_excel()`: reading Excel sheets.
- `pd.read_json()`, `pd.read_html()`, `pd.read_sql()`: reading other formats.
- Important arguments:
  - `sep`, `encoding`, `na_values`, `usecols`, `nrows`, `parse_dates`, `dtype`.
- Inspection methods:
  - `.head()`, `.info()`, `.describe()`, `.shape`, `.dtypes`.


### Importing CSV Files

A CSV file is a type of file where each line contains a single record, and all the columns are separated from each other via comma (_or others_).

To read a CSV file and load it into a Pandas dataframe, you can use the `pd.read_csv()` method.



In [17]:
iris_data = pd.read_csv("../../00_Datasets/iris.csv")
iris_data.head()

,sepal.length,sepal.width,petal.length,petal.width,variety
0,5.1,3.5,1.4,0.2,Setosa
1,4.9,3.0,1.4,0.2,Setosa
2,4.7,3.2,1.3,0.2,Setosa
3,4.6,3.1,1.5,0.2,Setosa
4,5.0,3.6,1.4,0.2,Setosa


In some cases, CSV files do not contain any header. In such
cases, the read_csv() method treats the first row of the CSV
file as the dataframe header.


To specify custom headers for your CSV files, you need to
pass the list of headers to the names attribute of the read_
csv() method, as shown in the script below. You can find the
“pima-indians-diabetes.csv” file in the Data folder of the book
resources.

In [18]:
iris_no_header = pd.read_csv("../../00_Datasets/iris_no_header.csv",
                             names=['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'class'])
iris_no_header.head()

,sepal_length,sepal_width,petal_length,petal_width,class
0,5.1,3.5,1.4,0.2,Setosa
1,4.9,3.0,1.4,0.2,Setosa
2,4.7,3.2,1.3,0.2,Setosa
3,4.6,3.1,1.5,0.2,Setosa
4,5.0,3.6,1.4,0.2,Setosa


### Importing Excel Files

To read an Excel file and load it into a Pandas dataframe, you can use the `pd.read_excel()` method.

ps: `openpyxl` library is required to read Excel files with pandas.

In [19]:
# !pip install openpyxl

In [20]:
excel_data = pd.read_excel("../../00_Datasets/supply_sales.xlsx")
excel_data.head()

,OrderDate,Region,Rep,Item,Units,UnitCost,Total
0,2024-06-01 00:00:00,East,Jones,Pencil,95,1.99,189.05
1,1/23/2024,Central,Kivell,Binder,50,19.99,999.50
2,2024-09-02 00:00:00,Central,Jardine,Pencil,36,4.99,179.64
3,2/26/2024,Central,Gill,Pen,27,19.99,539.73
4,3/15/2024,West,Sorvino,Pencil,56,2.99,167.44


`shet_name` parameter is used to specify the name of the sheet you want to read from the Excel file.

In [21]:
excel_data = pd.read_excel("../../00_Datasets/supply_sales.xlsx", sheet_name='Sheet1')
excel_data.head()

,OrderDate,Region,Rep,Item,Units,UnitCost,Total
0,2024-06-01 00:00:00,East,Jones,Pencil,95,1.99,189.05
1,1/23/2024,Central,Kivell,Binder,50,19.99,999.50
2,2024-09-02 00:00:00,Central,Jardine,Pencil,36,4.99,179.64
3,2/26/2024,Central,Gill,Pen,27,19.99,539.73
4,3/15/2024,West,Sorvino,Pencil,56,2.99,167.44


### Importing JSON Files

To read a JSON file and load it into a Pandas dataframe, you can use the `pd.read_json()` method.

`orient` parameter is used to specify the format of the JSON file.
- `'index'`: Dictionary like {index -> {column -> value}}
- `'columns'`: Dictionary like {column -> {index -> value}}
- `'records'`: List like [{column -> value}, ... , {column -> value}]


In [22]:
movies_data = pd.read_json("../../00_Datasets/movies_reviews.json", orient='index')
movies_data

,title,year,imdb_id,letterboxd_url,letterboxd_score,letterboxd_plot,genres,countries,languages,reviews
https://letterboxd.com/film/kneecap/,Kneecap,2024,tt27367464,https://letterboxd.com/film/kneecap/,4.1,When fate brings Belfast teacher JJ into the o...,"[Comedy, Drama, Music, Humanity and the world ...","[Ireland, UK]","[Irish, Irish, English]","[{'username': 'Meri', 'user_profile': 'https:/..."
https://letterboxd.com/film/the-girl-with-the-needle/,The Girl with the Needle,2024,tt10236164,https://letterboxd.com/film/the-girl-with-the-...,3.9,"Struggling to survive in post-WWI Copenhagen, ...","[Drama, History, Intense violence and sexual t...","[Denmark, Poland, Sweden]",[Danish],"[{'username': 'Aragonez94', 'user_profile': 'h..."
https://letterboxd.com/film/amelias-children/,Amelia’s Children,2023,tt26594993,https://letterboxd.com/film/amelias-children/,2.5,When Edward’s search for his biological family...,"[Mystery, Horror, Horror, the undead and monst...",[Portugal],"[Portuguese, English, Spanish, Portuguese]","[{'username': 'BeesKnees313', 'user_profile': ..."
https://letterboxd.com/film/john-vardar-vs-the-galaxy/,John Vardar vs the Galaxy,2024,tt9105364,https://letterboxd.com/film/john-vardar-vs-the...,NaN,One completely unexceptional man. One emotiona...,"[Animation, Adventure, Science Fiction, Comedy]","[Bulgaria, Croatia, Hungary, North Macedonia]","[Macedonian, Croatian, Bulgarian, English, Mac...","[{'username': 'theatre13', 'user_profile': 'ht..."



## Exercício Prático

1. Realize a leitura do ficheiro CSV `titanic.csv` disponível na pasta `files`.
2. Utilizando o método `.info()`, verifique o número de entradas, colunas e tipos de dados.
    - Quantas colunas existem?
    - Que colunas têm valores nulos?
3. Utilize o método `.describe()` para obter estatísticas descritivas das colunas numéricas.
    - Qual é a média da idade dos passageiros?
    - Qual é o valor máximo da coluna `Fare`?
    - Quantos passageiros sobreviveram (coluna `Survived`)? -----



In [23]:
titanic_data = pd.read_csv("../../00_Datasets/titanic.csv")
# titanic_data.info()
titanic_data.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


# Manipulating Dataframes with Pandas

## Handling Missing Values

Missing values, as the name suggests, are those observations in the dataset that doesn’t contain any value.

Missing values can change the data patterns, and, therefore, it is extremely important to understand why missing values occur in the
dataset and how to handle them.

- We can identify missing values in a Pandas dataframe using the `isna()` method. This method returns a dataframe of the same shape as the original dataframe, with boolean values indicating whether each value is missing (NaN) or not.
- `notna()` method can also be used to identify non-missing values.


### Handling Missing Numerical Values

There are several ways to handle missing numerical values in a dataset. Some of the most common methods include:
1. **Removing Rows with Missing Values**: You can remove rows that contain missing values using the `dropna()` method.
2. **Filling Missing Values with a Constant**: You can fill missing values with a constant value using the `fillna()` method.
3. **Filling Missing Values with Mean/Median/Mode** (Imputation): You can fill missing values with the mean, median, or mode of the column using the `fillna()` method.
4. **Using Interpolation**: You can use interpolation to estimate missing values based on other values in the column.

In [24]:
titanic = pd.read_csv("../../00_Datasets/titanic.csv")
print(f'Titanic dataset rows: {len(titanic)}')

Titanic dataset rows: 891


In [25]:
# percentage of null values per column
print('Percentage of null values per column:')
print(titanic.isna().mean())

Percentage of null values per column:
PassengerId    0.000000
Survived       0.000000
Pclass         0.000000
Name           0.000000
Sex            0.000000
Age            0.198653
SibSp          0.000000
Parch          0.000000
Ticket         0.000000
Fare           0.000000
Cabin          0.771044
Embarked       0.002245
dtype: float64


In [26]:
# Remove rows with missing values
titanic_dropped = titanic.dropna(subset=['Age'])
print(f'Titanic dataset rows without nulls values for col \'Age\': {len(titanic_dropped)}')

Titanic dataset rows without nulls values for col 'Age': 714


In [27]:
# Fill missing values with a constant
print('Filling missing values with a constant (28)')
print(f'Titanic null values for col Age {titanic['Age'].isna().sum()}')
titanic_filled_constant = titanic.copy()
titanic_filled_constant.fillna({'Age': 30}, inplace=True)
print(f'Titanic null values for col \'Age\' {titanic_filled_constant['Age'].isna().sum()}')


Filling missing values with a constant (28)
Titanic null values for col Age 177
Titanic null values for col 'Age' 0


In [28]:
# Fill missing values with mean/median/mode
print(f'Filling missing values with mean/median ({titanic['Age'].mean():.2f} / {titanic['Age'].median():.2f})')
titanic_filled_mean = titanic.copy()
titanic_filled_mean.fillna({'Age': titanic['Age'].median()}, inplace=True)
print(f'Titanic null values for col \'Age\' {titanic_filled_mean['Age'].isna().sum()}')


Filling missing values with mean/median (29.70 / 28.00)
Titanic null values for col 'Age' 0


### Handling Missing Categorical Values

Handling missing categorical values can be done using similar methods as numerical values, with some variations:
1. **Removing Rows with Missing Values**: You can remove rows that contain missing values using the `dropna()` method.
2. **Filling Missing Values with a Constant**: You can fill missing values with a constant value (e.g., 'Unknown') using the `fillna()` method.
3. **Filling Missing Values with Mode**: You can fill missing values with the mode (most frequent value) of the column using the `fillna()` method.
4. **Using Forward Fill or Backward Fill**: You can use forward fill (`ffill`) or backward fill (`bfill`) to propagate the next or previous values to fill missing values. (useful for time series data)


_For this example, we will use the `Embarked` column from the Titanic dataset, which contains categorical values representing the port of embarkation. Possible values are 'C' (Cherbourg), 'Q' (Queenstown), and 'S' (Southampton)._


In [29]:
titanic = pd.read_csv("../../00_Datasets/titanic.csv")
print(f'Titanic dataset rows: {len(titanic)}\n')

Titanic dataset rows: 891



In [30]:
print('Percentage of null values per column:')
print(titanic.isnull().mean(), '\n')

Percentage of null values per column:
PassengerId    0.000000
Survived       0.000000
Pclass         0.000000
Name           0.000000
Sex            0.000000
Age            0.198653
SibSp          0.000000
Parch          0.000000
Ticket         0.000000
Fare           0.000000
Cabin          0.771044
Embarked       0.002245
dtype: float64 



In [31]:
# Remove rows with missing values
titanic_dropped = titanic.dropna(subset=['Embarked'])
print(f'Titanic dataset rows without nulls values for col \'Embarked\': {len(titanic_dropped)}\n')

Titanic dataset rows without nulls values for col 'Embarked': 889



In [32]:
# Fill missing values with a constant
print('Filling missing values with a constant (Unknown)')
print(f'Titanic null values for col \'Embarked\' {titanic['Embarked'].isna().sum()}')
titanic_filled_constant = titanic.copy()
titanic_filled_constant.fillna({'Embarked': 'Unknown'}, inplace=True)
print(f'Titanic null values for col \'Embarked\' after fill {titanic_filled_constant['Embarked'].isna().sum()}')

Filling missing values with a constant (Unknown)
Titanic null values for col 'Embarked' 2
Titanic null values for col 'Embarked' after fill 0


In [33]:
# Fill missing values with mode
mode_embarked = titanic['Embarked'].mode()[0]
# print(titanic['Embarked'].mode()[0]) #['S']
print(f'\nFilling missing values with mode ({mode_embarked})')
titanic_filled_mode = titanic.copy()
titanic_filled_mode.fillna({'Embarked': mode_embarked}, inplace=True)
print(f'Titanic null values for col \'Embarked\' {titanic_filled_mode['Embarked'].isna().sum()}')




Filling missing values with mode (S)
Titanic null values for col 'Embarked' 0


## Duplicates and Type Conversion

Sometimes, datasets may contain duplicate records or values that are not in the desired format. Pandas provides methods to handle these issues effectively.



### Handling Duplicates
- **Identifying Duplicates**: You can identify duplicate rows in a Pandas dataframe using the `duplicated()` method. This method returns a boolean Series indicating whether each row is a duplicate of a previous row.
- **Removing Duplicates**: You can remove duplicate rows using the `drop_duplicates()` method. You can specify which columns to consider for identifying duplicates using the `subset` parameter.


In [34]:
data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'Alice', 'Bob'],
    'Age': [25, 30, 35, 25, 31],
    'City': ['New York', 'Los Angeles', 'Chicago', 'New York', 'Los Angeles']
}
df = pd.DataFrame(data)

print("Original DataFrame:")
display(df)

# Identify duplicates
duplicates = df.duplicated()
# print("\nDuplicate Rows:")
display(df[duplicates])
#
# Remove duplicates
df_no_duplicates = df.drop_duplicates()
print("\nDataFrame after removing duplicates:")
display(df_no_duplicates)

Original DataFrame:


,Name,Age,City
0,Alice,25,New York
1,Bob,30,Los Angeles
2,Charlie,35,Chicago
3,Alice,25,New York
4,Bob,31,Los Angeles


,Name,Age,City
3,Alice,25,New York



DataFrame after removing duplicates:


,Name,Age,City
0,Alice,25,New York
1,Bob,30,Los Angeles
2,Charlie,35,Chicago
4,Bob,31,Los Angeles


- Duplicates can also be dropped based on specific columns by passing a list of column names to the `subset` parameter of the `drop_duplicates()` method.

In [35]:
df_no_duplicates = df.drop_duplicates(subset=['Name', 'City'])
print("\nDataFrame after removing duplicates based on 'Name' and 'City':")
display(df_no_duplicates)



DataFrame after removing duplicates based on 'Name' and 'City':


,Name,Age,City
0,Alice,25,New York
1,Bob,30,Los Angeles
2,Charlie,35,Chicago


### Type Conversion
- **Changing Data Types**: You can change the data type of a column using the `astype()` method. This is useful when you need to convert a column to a different type for analysis or visualization.
- **Common Conversions**: Common conversions include converting strings to numeric types, converting numeric types to categorical types, and converting dates to datetime types

In [36]:
data = {
    'Name': ['Alice', 'Bob', 'Charlie'],
    'Age': ['25', '30', '35'],
    'Join_Date': ['2020-01-15', '2019-06-23', '2021-03-10']
}

df = pd.DataFrame(data)
print("Original DataFrame:")
display(df)
print(df.info())

# Convert 'Age' from string to integer
df['Age'] = df['Age'].astype(int)
print("\nDataFrame after converting 'Age' to integer:")
print(df.info())
#
# Convert 'Join_Date' from string to datetime
df['Join_Date'] = pd.to_datetime(df['Join_Date'])
print("\nDataFrame after converting 'Join_Date' to datetime:")
print(df.info())
df

Original DataFrame:


,Name,Age,Join_Date
0,Alice,25,2020-01-15
1,Bob,30,2019-06-23
2,Charlie,35,2021-03-10


<class 'pandas.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   Name       3 non-null      str  
 1   Age        3 non-null      str  
 2   Join_Date  3 non-null      str  
dtypes: str(3)
memory usage: 204.0 bytes
None

DataFrame after converting 'Age' to integer:
<class 'pandas.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   Name       3 non-null      str  
 1   Age        3 non-null      int64
 2   Join_Date  3 non-null      str  
dtypes: int64(1), str(2)
memory usage: 204.0 bytes
None

DataFrame after converting 'Join_Date' to datetime:
<class 'pandas.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Name       3 non-null      str           
 1   Age   

,Name,Age,Join_Date
0,Alice,25,2020-01-15
1,Bob,30,2019-06-23
2,Charlie,35,2021-03-10


Datetime allows us to work with dates and times in a more efficient way. We can also specify the format of the date string while converting it to datetime.

The common format codes are:

- Y - Year with century as a decimal number (ex: 1999, 2003, 2024)
- y - Year without century as a zero-padded decimal number (ex: 99, 03, 24)
- m - Month as a zero-padded decimal number (01 to 12)
- B - Full month name (ex: January, February, etc.)
- b - Abbreviated month name (ex: Jan, Feb, etc.)
- d - Day of the month as a zero-padded decimal number (01 to 31)
- H - Hour (24-hour clock) as a zero-padded decimal number (00 to 23)
- I - Hour (12-hour clock) as a zero-padded decimal number (01 to 12)
- M - Minute as a zero-padded decimal number (00 to 59)
- S - Second as a zero-padded decimal number (00 to 59)


In [37]:
df['Join_Date'] = pd.to_datetime(df['Join_Date'], format='%Y-%m-%d')
print("\nDataFrame after converting 'Join_Date' to datetime with specific format:")
print(df.info())
display(df)


DataFrame after converting 'Join_Date' to datetime with specific format:
<class 'pandas.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Name       3 non-null      str           
 1   Age        3 non-null      int64         
 2   Join_Date  3 non-null      datetime64[us]
dtypes: datetime64[us](1), int64(1), str(1)
memory usage: 204.0 bytes
None


,Name,Age,Join_Date
0,Alice,25,2020-01-15
1,Bob,30,2019-06-23
2,Charlie,35,2021-03-10


## Exercício Prático

1. Lê o ficheiro CSV arbnb lisbon.csv localizado na pasta files.
2. Utiliza `df.info()` para verificar:
    - o número de registos (entries) **24264**
    - os tipos de dados de cada coluna.
    - Quantas colunas existem? **79**
    - Quais são as colunas que contêm valores nulos?
    - Qual a coluna com mais valores nulos?
3. Usa `df.describe()` para obter estatísticas descritivas das colunas numéricas.
    - Qual é a média da coluna `price`?
    - Qual é o valor máximo da coluna `price`?
    - Quantos anúncios possuem pelo menos uma review?
4. Identifica e remove quaisquer registos duplicados no dataframe.
    - Existem duplicados no dataframe? Quantos?
5. Remover duplicados utilizando diferentes estratégias de `drop_duplicates()` e comparar:
    - Remove duplicados com base nos subsets ['name'] e ['name', 'latitude', 'longitude'].
    - Remove duplicados com base em todas as colunas.
        - Quantos registos permanecem após cada remoção de duplicados?
        - Recalcule o preço médio. (o que acontece ao preço médio após remover duplicados?)
                -

~~lisbon['price'] = lisbon['price'].str.replace('$', '').str.replace(',', '').astype(float)~~


In [38]:
lisbon = pd.read_csv("../../00_Datasets/lisbon.csv")
lisbon['price'] = lisbon['price'].str.replace('$', '').str.replace(',', '').astype(float)
# lisbon.info()
display(lisbon.describe()[['id', 'price']])

duplicados1 = lisbon.duplicated(subset=['name'])
duplicados2 = lisbon.duplicated(subset=['name', 'latitude', 'longitude'])
print(duplicados1.sum() / len(lisbon), duplicados2.sum() / len(lisbon))

no_duplicates1 = lisbon.drop_duplicates(subset=['name'])
no_duplicates2 = lisbon.drop_duplicates(subset=['name', 'latitude', 'longitude'])
display(no_duplicates1.describe()[['id', 'price']])  # 155.22
no_duplicates2.describe()[['id', 'price']]  # 156.92


,id,price
count,2.426400e+04,21079.000000
mean,5.075646e+17,156.624982
std,5.357909e+17,576.065191
min,6.499000e+03,8.000000
25%,2.442599e+07,60.000000
50%,5.423861e+07,87.000000
75%,1.044822e+18,132.000000
max,1.371345e+18,20000.000000


0.032847016155621495 0.00511045169798879


,id,price
count,2.346700e+04,20451.000000
mean,4.929719e+17,155.228742
std,5.328480e+17,573.601098
min,6.499000e+03,8.000000
25%,2.393938e+07,60.000000
50%,5.299758e+07,88.000000
75%,1.024657e+18,133.000000
max,1.371345e+18,20000.000000


,id,price
count,2.414000e+04,20991.000000
mean,5.050377e+17,156.922205
std,5.353116e+17,577.227082
min,6.499000e+03,8.000000
25%,2.434825e+07,60.000000
50%,5.393252e+07,87.000000
75%,1.040754e+18,133.000000
max,1.371345e+18,20000.000000


## **Revisão parte 1**

- Pandas Series e DataFrames
- Importação de dados (CSV, Excel, JSON)
- `info()` and `describe()` methods
- Tratamento de valores em falta (missing values), que abordagens?
- Tratamento de duplicados (o que é um duplicado) e conversão de tipos de dados


## Selecting Data Using Indexing and Slicing

Indexing refers to fetching data using index or column information of a Pandas dataframe. Slicing, on the other hand, refers to slicing a Pandas dataframe using indexing techniques.


### Selecting Data Using Brackets []

One of the simplest ways to select data from various columns
is by using square brackets. To get column data in the form of
a series from a Pandas dataframe, you need to pass the
column name inside square brackets that follow the Pandas
dataframe name.

Eg. `df['column_name']`

In [39]:
titanic_data = pd.read_csv("../../00_Datasets/titanic.csv")
# titanic_data.columns
print(titanic_data["Sex"])
type(titanic_data["Sex"])

0        male
1      female
2      female
3      female
4        male
        ...  
886      male
887    female
888    female
889      male
890      male
Name: Sex, Length: 891, dtype: str


pandas.Series

You can select multiple columns by passing a list of column
names inside a string to the square brackets. You will then get
a Pandas dataframe with the specified columns, as shown
below.

In [40]:
titanic_data[['Name', 'Age']]

,Name,Age
0,"Braund, Mr. Owen Harris",22.0
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0
2,"Heikkinen, Miss. Laina",26.0
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0
4,"Allen, Mr. William Henry",35.0
...,...,...
886,"Montvila, Rev. Juozas",27.0
887,"Graham, Miss. Margaret Edith",19.0
888,"Johnston, Miss. Catherine Helen ""Carrie""",NaN
889,"Behr, Mr. Karl Howell",26.0


### Indexing and Slicing Using loc Function

The `loc` function from the Pandas dataframe can also be use to filter records in the Pandas dataframe.

The `loc` function is used to select data based on labels or boolean arrays. It allows you to select rows and columns by their labels.

In [41]:
scores = [
    {'Subject': 'Mathematics', 'Score': 85, 'Grade': 'B', 'Remarks': 'Good', },
    {'Subject': 'History', 'Score': 98, 'Grade': 'A', 'Remarks': 'Excellent'},
    {'Subject': 'English', 'Score': 76, 'Grade': 'C', 'Remarks': 'Fair'},
    {'Subject': 'Science', 'Score': 72, 'Grade': 'C', 'Remarks': 'Fair'},
    {'Subject': 'Arts', 'Score': 95, 'Grade': 'A', 'Remarks': 'Excellent'},
]
my_df = pd.DataFrame(scores)
my_df.head()

,Subject,Score,Grade,Remarks
0,Mathematics,85,B,Good
1,History,98,A,Excellent
2,English,76,C,Fair
3,Science,72,C,Fair
4,Arts,95,A,Excellent


In [42]:
print(my_df.loc[2])
type(my_df.loc[2])

Subject    English
Score           76
Grade            C
Remarks       Fair
Name: 2, dtype: object


pandas.Series

In [43]:
my_df.loc[2:4]

,Subject,Score,Grade,Remarks
2,English,76,C,Fair
3,Science,72,C,Fair
4,Arts,95,A,Excellent


Along with filtering rows, you can also specify which columns to filter with the loc function.

The following script filters the values in columns `Grade` and `Score` in the rows from index _2 to 4_.

In [44]:
my_df.loc[2:4, ["Grade", "Score"]]

,Grade,Score
2,C,76
3,C,72
4,A,95


### Indexing and Slicing Using iloc Function

The `iloc` function from the Pandas dataframe can also be use to filter records in the Pandas dataframe.

The `iloc` function is used to select data based on integer position. It allows you to select rows and columns by their integer index.




In [45]:
scores = [
    {'Subject': 'Mathematics', 'Score': 85, 'Grade': 'B', 'Remarks': 'Good', },
    {'Subject': 'History', 'Score': 98, 'Grade': 'A', 'Remarks': 'Excellent'},
    {'Subject': 'English', 'Score': 76, 'Grade': 'C', 'Remarks': 'Fair'},
    {'Subject': 'Science', 'Score': 72, 'Grade': 'C', 'Remarks': 'Fair'},
    {'Subject': 'Arts', 'Score': 95, 'Grade': 'A', 'Remarks': 'Excellent'},
]
my_df = pd.DataFrame(scores)
my_df.head()

my_df.iloc[3]

Subject    Science
Score           72
Grade            C
Remarks       Fair
Name: 3, dtype: object

If you want to select records from a single column as a
dataframe, you need to specify the index inside the square
brackets and then those square brackets inside the square
brackets that follow the iloc function


🤷🏼‍♂️

In [46]:
my_df.iloc[[3]]

,Subject,Score,Grade,Remarks
3,Science,72,C,Fair


You can also pass a range of indexes.

In [47]:
my_df.iloc[3:5]

,Subject,Score,Grade,Remarks
3,Science,72,C,Fair
4,Arts,95,A,Excellent


In addition to specifying indexes, you can also pass column numbers (starting from 0) to the `iloc` method

In [48]:
# my_df.iloc[[0, 2, 3], [0, 1]]

#or

my_df.iloc[2:4, 0:2]

,Subject,Score
2,English,76
3,Science,72


### overview

| Type | Notes                                                                                                                                                                                                |
|------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| df[column] | Select single column or sequence of columns from the DataFrame|
| df.loc[rows] | Select single row or subset of rows from the DataFrame by label                                                                                                                                      |
| df.loc[:, cols] | Select single column or subset of columns by label                                                                                                                                                   |
| df.loc[rows, cols] | Select both row(s) and column(s) by label                                                                                                                                                            |
| df.iloc[rows] | Select single row or subset of rows from the DataFrame by integer position                                                                                                                           |
| df.iloc[:, cols] | Select single column or subset of columns by integer position                                                                                                                                        |
| df.iloc[rows, cols] | Select both row(s) and column(s) by integer position                                                                                                                                                 |
| df.at[row, col] | Select a single scalar value by row and column label                                                                                                                                                 |
| df.iat[row, col] | Select a single scalar value by row and column position (integers)                                                                                                                                   |
| reindex method | Select either rows or columns by labels                                                                                                                                                              |


## Filtering Rows



###  Filtering Based on a Single Condition

You can filter rows in a Pandas dataframe based on a single condition by using boolean indexing.

In [49]:
data = {'Name': ['Alice', 'Bob', 'Charlie', 'David'],
        'Age': [25, 30, 35, 40],
        'City': ['New York', 'Chicago', 'New York', 'San Francisco']}

df = pd.DataFrame(data)
df
# boolean mask
df['Age'] > 30

0    False
1    False
2     True
3     True
Name: Age, dtype: bool

In [50]:
# Filtering rows where Age is greater than 30
filtered_df = df[df['Age'] > 30]
display(filtered_df)

,Name,Age,City
2,Charlie,35,New York
3,David,40,San Francisco


### Filtering Using Multiple Conditions (`&`, `|`, `~`)

You can filter rows in a Pandas dataframe based on multiple conditions by combining boolean expressions using logical operators:
- `&` (AND): Both conditions must be true.
- `|` (OR): At least one condition must be true.
- `~` (NOT): Negates the condition.


In [51]:
mask = (df['Age'] > 30) & (df['City'] == 'New York')

filtered_df = df[mask]
display(filtered_df)

,Name,Age,City
2,Charlie,35,New York


In [52]:
(df['Age'] > 30) | (df['City'] == 'New York')
filtered_df = df[(df['Age'] > 30) | (df['City'] == 'New York')]
display(filtered_df)

,Name,Age,City
0,Alice,25,New York
2,Charlie,35,New York
3,David,40,San Francisco


In [53]:
# Exclude rows where City is 'Chicago'

filtered_df = df[~(df['City'] == 'Chicago')]
display(filtered_df)

,Name,Age,City
0,Alice,25,New York
2,Charlie,35,New York
3,David,40,San Francisco


### Filtering Using .query() Method

The `query()` method provides a more readable way to filter rows in a Pandas dataframe using a query string.


In [54]:
filtered_df = df.query("Age > 30 and City == 'New York'")
display(filtered_df)

,Name,Age,City
2,Charlie,35,New York


### Advanced Filtering Techniques in Pandas

- Filtering with `.isin()`: Check if values in a column are present in a list of values.
- Filtering with `.between()`: Check if values in a column fall within a specified range.
- Filtering with **_string_** methods: Use string methods like `.str.contains()`, `.str.startswith()`, and `.str.endswith()` for text-based filtering.

In [55]:
# isin()
# df['City'].isin(['New York', 'Chicago', 'Lisboa'])
filtered_df = df[df['City'].isin(['New York', 'Chicago'])]
display(filtered_df)

,Name,Age,City
0,Alice,25,New York
1,Bob,30,Chicago
2,Charlie,35,New York


In [56]:
# between()
filtered_df = df[df['Age'].between(30, 40)]
display(filtered_df)

,Name,Age,City
1,Bob,30,Chicago
2,Charlie,35,New York
3,David,40,San Francisco


In [57]:
# string methods
## contains()
# df['Name'].info()
display(df)
# df['Name'].str.contains('A', case=True)
filtered_df = df[df['Name'].str.contains('e', case=False)]
display(filtered_df)
#
## startswith()
filtered_df = df[df['Name'].str.startswith('A')]
display(filtered_df)

#7


,Name,Age,City
0,Alice,25,New York
1,Bob,30,Chicago
2,Charlie,35,New York
3,David,40,San Francisco


,Name,Age,City
0,Alice,25,New York
2,Charlie,35,New York


,Name,Age,City
0,Alice,25,New York


##  Exercício Prático

Usando o dataset `lisbon.csv`, realiza as seguintes operações:

1. Seleciona a coluna `price` utilizando apenas `[]`.
2. Seleciona as colunas `name`, `price` e `neighbourhood` usando `df[[]]`.
3. Usa `.loc` para selecionar:
    - todas as linhas,
    - apenas as colunas `host_name` e `room_type`.

4. Usa `.loc` para selecionar as linhas com índice de 10 a 20 (inclusive) e as colunas `name` e `price`.
5. Usa `.iloc` para selecionar as **5 primeiras linhas** e **as 3 primeiras colunas**.

6. Usa `.iloc` para selecionar:
    - as linhas nas posições 100 a 150,
    - as colunas nas posições 0, 2 e 5.

7. Filtra as linhas onde `price > 100`.
8. Filtra as linhas onde `room_type == "Entire home/apt"`.

9. Filtra as linhas onde `number_of_reviews == 0`.
10. Filtra as linhas onde:
    - `price > 80` **e** `number_of_reviews > 10`.
11. Filtra as linhas onde:
    - `room_type == "Private room"` **ou** `price < 50`.
12. Filtra anúncios onde:
    - `neighbourhood == "Alfama"` **e**  `availability_365 > 200`.
13. Usa `.query()` para obter todas as linhas onde `price < 30`.
14. Usa `.query()` para filtrar:
    - `room_type == "Shared room"`  **e** `number_of_reviews > 5`.
15. Usa `.query()` para selecionar anúncios com:
    - ``price >= 50 and price <= 120``
16. Usa `isin()` para selecionar apenas os bairros:
    - Bairro Alto
    - Alfama
    - Arroios

    (coluna: `neighbourhood`)
17. Usa `isin()` para selecionar apenas os `room_type`:
    - `"Entire home/apt"`
    - `"Private room"`

18. Selecione anúncios com `price` entre 40 e 100 (inclusive).
19. Filtre anuncios com `number_of_reviews` entre 10 e 50.




In [58]:
df = pd.read_csv("../../00_Datasets/lisbon.csv")
df['price'] = df['price'].str.replace('$', '').str.replace(',', '').astype(float)

In [59]:
# df = pd.read_csv("../files/lisbon.csv")

#1
# print(df['price'])

# 2
# display(df[['name', 'price', 'neighbourhood']])

#3
# df.loc[1:5 ,['host_name', 'room_type']]

#4
# df.loc[10:20, ['name', 'price']]

#5
# df.loc[:4]
# df.iloc[:5, :3]

#6
# df.iloc[100:151, [0,2,5]]

#7
# df['price_wo_dollar'] = df['price'].str.replace('$', '')
# df['price_wo_dollar']
# df['price_wo_dollar_and_comma'] = df['price_wo_dollar'].str.replace(',', '')
# df['price'] = df['price_wo_dollar_and_comma'].astype(float)
# df['price'] = df['price'].str.replace('$', '').str.replace(',', '').astype(float)
# print(f'7. Dataset não filtrado {len(df)}')
# df_filtered = df[df['price'] > 100]
# print(f'7. Dataset após filtro de preço > 100 {len(df_filtered)}')


# 12
# df['availability_365']
# df['neighbourhood'].unique()
# df_filtered = df[(df['availability_365'] > 200) & (df['neighbourhood'].str.contains('Alfama', case=False))]
# df_filtered[['availability_365', 'neighbourhood']]
#

#16
df['neighbourhood'].isna().mean()
df_filtered = df.dropna(subset=['neighbourhood'])
df_filtered = df_filtered[df_filtered['neighbourhood'].str.contains('alfama|bairro Alto|Arroios', case=False)]
# df_filtered[['neighbourhood']]

#17
# df_filtered[df_filtered['room_type'].isin(['Entire home/apt', 'Private room'])]
# df_filtered[['room_type']]

# # 18
# df_filtered = df[df['price'].between(40, 100)]

## Sorting DataFrames

To sort the Pandas dataframe, you can use the `sort_values()` function of the Pandas dataframe. The list of columns used for sorting needs to be passed to the by attribute of the `sort_values()` method.


In [60]:
titanic_data = pd.read_csv("../../00_Datasets/titanic.csv")
display(titanic_data.head())

age_sorted_data = titanic_data.sort_values(by=['Age'])
age_sorted_data.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
803,804,1,3,"Thomas, Master. Assad Alexander",male,0.42,0,1,2625,8.5167,NaN,C
755,756,1,2,"Hamalainen, Master. Viljo",male,0.67,1,1,250649,14.5000,NaN,S
644,645,1,3,"Baclini, Miss. Eugenie",female,0.75,2,1,2666,19.2583,NaN,C
469,470,1,3,"Baclini, Miss. Helene Barbara",female,0.75,2,1,2666,19.2583,NaN,C
78,79,1,2,"Caldwell, Master. Alden Gates",male,0.83,0,2,248738,29.0000,NaN,S


To sort by descending order, you need to pass False as the value for the `ascending` attribute.


In [61]:
age_sorted_data = titanic_data.sort_values(by=['Age', ], ascending=False)
age_sorted_data.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
630,631,1,1,"Barkworth, Mr. Algernon Henry Wilson",male,80.0,0,0,27042,30.0000,A23,S
851,852,0,3,"Svensson, Mr. Johan",male,74.0,0,0,347060,7.7750,NaN,S
493,494,0,1,"Artagaveytia, Mr. Ramon",male,71.0,0,0,PC 17609,49.5042,NaN,C
96,97,0,1,"Goldschmidt, Mr. George B",male,71.0,0,0,PC 17754,34.6542,A5,C
116,117,0,3,"Connors, Mr. Patrick",male,70.5,0,0,370369,7.7500,NaN,Q


## Pandas `Unique` and `Count` Functions

- `unique()` function in Pandas is used to find the unique values in a column of a dataframe. It returns an array of unique values present in the specified column.
- `nunique()` function in Pandas is used to count the number of unique values in a column of a dataframe. It returns an integer representing the count of unique values present in the specified column.
- `count()` function in Pandas is used to count the number of non-null values in a column of a dataframe. It returns an integer representing the count of non-null values present in the specified column.
- `value_counts()` function in Pandas is used to count the occurrences of each unique value in a column of a dataframe. It returns a Series containing the counts of unique values, sorted in descending order by default.

In [62]:
#unique()
titanic_data = pd.read_csv("../../00_Datasets/titanic.csv")
titanic_data
# unique_classes = titanic_data['Pclass'].unique()
# print("Unique classes in 'Pclass' column:", unique_classes)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


In [63]:
#nunique()
num_unique_classes = titanic_data['Pclass'].nunique()
print("Number of unique classes in 'Pclass' column:", num_unique_classes)

Number of unique classes in 'Pclass' column: 3


In [64]:
#count()
num_non_null = titanic_data.count()
num_non_null

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

In [65]:
# value_counts()
class_counts = titanic_data['Pclass'].value_counts()
print("Value counts for 'Pclass' column:")
print(class_counts)

Value counts for 'Pclass' column:
Pclass
3    491
1    216
2    184
Name: count, dtype: int64


## Exercício Prático

1. Ordena o dataframe pelo preço (`price`) em ordem ascendente.
2. Ordena o dataframe pelo número de reviews (`number_of_reviews`) em ordem descendente.
3. Ordena o dataframe primeiro pelo bairro (`number_of_reviews`) em ordem ascendente e depois pelo preço (`price`) em ordem descendente. (Mostra os ultimos 10 registos)
4. Mostra e conta o número de valores únicos na coluna `room_type`.
5. Conta o número de valores não nulos em cada coluna do dataframe.
6. Conta o número de anúncios por tipo de quarto (`room_type`).

In [66]:
# df['price'].info()

#3
# df.sort_values(by= ['price', 'number_of_reviews'], ascending=[True, False])[['number_of_reviews', 'price']]


#6
# df['room_type'].value_counts()

print(df['room_type'].value_counts())

df_gb_room_type = df.groupby('room_type').agg({'price': 'count'})
display(df_gb_room_type)

room_type
Entire home/apt    18084
Private room        5847
Hotel room           220
Shared room          113
Name: count, dtype: int64


,price
room_type,
Entire home/apt,16264
Hotel room,144
Private room,4581
Shared room,90


# Data Grouping, Aggregation

- `.groupby()` para agrupar por categorias.
- Agregações: `.sum()`, `.mean()`, `.count()`, `.agg()`.


## Grouping Data with GroupBy

To group data by a column value, you can use the `groupby()` function of the Pandas dataframe. You need to pass the column as the parameter value to the `groupby()` function.


In [67]:
titanic_gbclass = titanic_data.groupby(by="Embarked")
print(titanic_gbclass)

The above output shows that the `groupby() `function returns
the DataFrameGroupBy object. You can use various attributes
and functions of this object to get various information about
different groups.


In [68]:
print(titanic_gbclass.ngroups)
print(titanic_gbclass.size())
# print(titanic_gbclass.groups)

3
Embarked
C    168
Q     77
S    644
dtype: int64


In [69]:
#get a dataframe that contains records belonging to a subgroup using the get_group() function

titanic_second_class = titanic_gbclass.get_group('S')
titanic_second_class

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
...,...,...,...,...,...,...,...,...,...,...,...,...
883,884,0,2,"Banfield, Mr. Frederick James",male,28.0,0,0,C.A./SOTON 34068,10.5000,NaN,S
884,885,0,3,"Sutehall, Mr. Henry Jr",male,25.0,0,0,SOTON/OQ 392076,7.0500,NaN,S
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S


The `DataFrameByGroup` object can also be used to perform `aggregate` functions

In [70]:
titanic_gbclass['Age'].median()

Embarked
C    29.0
Q    27.0
S    28.0
Name: Age, dtype: float64

Similarly, you can also get information based on various aggregate functions bypassing the list of functions to the `agg()` method

In [71]:
titanic_gbclass['Fare'].agg(['max', 'min', 'count', 'median', 'mean'])

,max,min,count,median,mean
Embarked,,,,,
C,512.3292,4.0125,168,29.70,59.954144
Q,90.0000,6.7500,77,7.75,13.276030
S,263.0000,0.0000,644,13.00,27.079812


In [72]:
# more complex aggregation example

titanic_data.groupby(by="Embarked").agg({
    'Name': 'count',
    'Age': 'mean',
    'Fare': ['sum', 'mean', 'std'],
    'Survived': 'mean'
})

Name        Age        Fare                        Survived
         count       mean         sum       mean        std      mean
Embarked                                                             
C          168  30.814769  10072.2962  59.954144  83.912994  0.553571
Q           77  28.089286   1022.2543  13.276030  14.188047  0.389610
S          644  29.445397  17439.3988  27.079812  35.887993  0.336957

In [73]:
titanic_data.groupby(by=['Embarked', 'Sex']).agg({
    'Fare': ['mean', 'max'],
    'SibSp': 'mean',
})

Fare               SibSp
                      mean       max      mean
Embarked Sex                                  
C        female  75.169805  512.3292  0.547945
         male    48.262109  512.3292  0.263158
Q        female  12.634958   90.0000  0.222222
         male    13.838922   90.0000  0.609756
S        female  38.740929  263.0000  0.837438
         male    21.711996  263.0000  0.448980

#  Concatenating and Merging Data


## Concatenating Data

Oftentimes, you need to concatenate or join multiple Pandas
dataframes horizontally or vertically. So, let’s first see how to
concatenate or join Pandas dataframes vertically or in row
order.

In [74]:
titanic_data = pd.read_csv("../../00_Datasets/titanic.csv")

titanic_pclass1_data = titanic_data[titanic_data["Pclass"] == 1]
print(titanic_pclass1_data.shape)

titanic_pclass2_data = titanic_data[titanic_data["Pclass"] == 2]
print(titanic_pclass2_data.shape)


(216, 12)
(184, 12)


The output shows that both the newly created dataframes
have 15 columns. It is important to mention that while
concatenating data vertically, both the dataframes should
have an _**equal**_ number of columns

In [75]:
final_data = pd.concat([titanic_pclass1_data, titanic_pclass2_data])
final_data.shape

(400, 12)

### Concatenating Columns

To concatenate dataframes horizontally, make sure that the dataframes have an equal number of rows.

You will need to pass `1` as the value for the `axis` attribute.

In [76]:
df1 = final_data[:200]
print(df1.shape)

df2 = final_data[200:]
print(df2.shape)

final_data = pd.concat([df1, df2], axis=1)
display(final_data)



(200, 12)
(200, 12)


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,...,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
1,2.0,1.0,1.0,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1.0,0.0,PC 17599,71.2833,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,4.0,1.0,1.0,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1.0,0.0,113803,53.1000,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,7.0,0.0,1.0,"McCarthy, Mr. Timothy J",male,54.0,0.0,0.0,17463,51.8625,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
11,12.0,1.0,1.0,"Bonnell, Miss. Elizabeth",female,58.0,0.0,0.0,113783,26.5500,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
23,24.0,1.0,1.0,"Sloper, Mr. William Thompson",male,28.0,0.0,0.0,113788,35.5000,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
866,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2.0,"Duran y More, Miss. Asuncion",female,27.0,1.0,0.0,SC/PARIS 2149,13.8583,NaN,C
874,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2.0,"Abelson, Mrs. Samuel (Hannah Wizosky)",female,28.0,1.0,0.0,P/PP 3381,24.0000,NaN,C
880,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2.0,"Shelley, Mrs. William (Imanita Parrish Hall)",female,25.0,0.0,1.0,230433,26.0000,NaN,S
883,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2.0,"Banfield, Mr. Frederick James",male,28.0,0.0,0.0,C.A./SOTON 34068,10.5000,NaN,S


## Merging DataFrames

Merging dataframes is similar to joining tables in a database. You can merge two dataframes based on a common column or index.

In [77]:
data = {
    'EmployeeID': [1, 2, 3, 4],
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'Department': ['HR', 'Finance', 'IT', 'Marketing']
}

df1 = pd.DataFrame(data)

data = {
    'EmployeeID': [3, 4, 5, 6],
    'Salary': [70000, 80000, 60000, 90000]
}
df2 = pd.DataFrame(data)
display(df1)
display(df2)

,EmployeeID,Name,Department
0,1,Alice,HR
1,2,Bob,Finance
2,3,Charlie,IT
3,4,David,Marketing


,EmployeeID,Salary
0,3,70000
1,4,80000
2,5,60000
3,6,90000


In [78]:
## Inner Join
inner_merged_df = pd.merge(df1, df2, on='EmployeeID', how='inner')
print("Inner Join Result:")
display(inner_merged_df)

Inner Join Result:


,EmployeeID,Name,Department,Salary
0,3,Charlie,IT,70000
1,4,David,Marketing,80000


In [79]:
# Left Join
left_merged_df = pd.merge(df1, df2, on='EmployeeID', how='left')
print("Left Join Result:")
display(left_merged_df)

Left Join Result:


,EmployeeID,Name,Department,Salary
0,1,Alice,HR,NaN
1,2,Bob,Finance,NaN
2,3,Charlie,IT,70000.0
3,4,David,Marketing,80000.0


In [80]:
# Right Join

right_merged_df = pd.merge(df1, df2, on='EmployeeID', how='right')
print("Right Join Result:")
display(right_merged_df)

Right Join Result:


,EmployeeID,Name,Department,Salary
0,3,Charlie,IT,70000
1,4,David,Marketing,80000
2,5,NaN,NaN,60000
3,6,NaN,NaN,90000


In [81]:
# Full Outer Join
outer_merged_df = pd.merge(df1, df2, on='EmployeeID', how='outer')
print("Full Outer Join Result:")
display(outer_merged_df)

Full Outer Join Result:


,EmployeeID,Name,Department,Salary
0,1,Alice,HR,NaN
1,2,Bob,Finance,NaN
2,3,Charlie,IT,70000.0
3,4,David,Marketing,80000.0
4,5,NaN,NaN,60000.0
5,6,NaN,NaN,90000.0


## Exercício Prático

1. Lê o ficheiro CSV `titanic.csv` localizado na pasta `files`.
2. Agrupa os dados pelo campo `Pclass` e calcula a média da idade (`Age`) para cada classe.
3. Agrupa os dados pelo campo `Survived` e calcula a soma dos passageiros (`Name`, porque sabemos que possivelmente é um campo com valores unicos) para cada grupo.
4. Concatena os dados dos passageiros da primeira classe (`Pclass` = 1) e da segunda classe (`Pclass` = 2) verticalmente.
5. Cria dois dataframes:
    - df1: contendo as colunas `PassengerId`, `Name` e `Pclass`.
    - df2: contendo as colunas `PassengerId` e `Survived`.
6. Realiza uma junção interna (inner join) entre df1 e df2 com base na coluna `PassengerId`.


In [82]:
titanic_data = pd.read_csv("../../00_Datasets/titanic.csv")
titanic_data


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


In [83]:
#2
# gb_class = titanic_data.groupby('Pclass')['Age'].mean()
# display(gb_class)

#3
gb_survived = titanic_data.groupby('Survived')['Name'].count()
gb_survived

#4
titanic_pclass1_data = titanic_data[titanic_data["Pclass"] == 1]
titanic_pclass2_data = titanic_data[titanic_data["Pclass"] == 2]
print(titanic_pclass1_data.shape, titanic_pclass2_data.shape)

concatenated_data = pd.concat([titanic_pclass1_data, titanic_pclass2_data])
print(concatenated_data.shape)

#5
df1 = titanic_data[['PassengerId', 'Name', 'Pclass']]
display(df1)
df2 = titanic_data[['PassengerId', 'Survived']]
display(df2)

#6
inner_merged_df = pd.merge(df1, df2, on='PassengerId', how='inner')
inner_merged_df


(216, 12) (184, 12)
(400, 12)


,PassengerId,Name,Pclass
0,1,"Braund, Mr. Owen Harris",3
1,2,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1
2,3,"Heikkinen, Miss. Laina",3
3,4,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1
4,5,"Allen, Mr. William Henry",3
...,...,...,...
886,887,"Montvila, Rev. Juozas",2
887,888,"Graham, Miss. Margaret Edith",1
888,889,"Johnston, Miss. Catherine Helen ""Carrie""",3
889,890,"Behr, Mr. Karl Howell",1


,PassengerId,Survived
0,1,0
1,2,1
2,3,1
3,4,1
4,5,0
...,...,...
886,887,0
887,888,1
888,889,0
889,890,1


,PassengerId,Name,Pclass,Survived
0,1,"Braund, Mr. Owen Harris",3,0
1,2,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,1
2,3,"Heikkinen, Miss. Laina",3,1
3,4,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,1
4,5,"Allen, Mr. William Henry",3,0
...,...,...,...,...
886,887,"Montvila, Rev. Juozas",2,0
887,888,"Graham, Miss. Margaret Edith",1,1
888,889,"Johnston, Miss. Catherine Helen ""Carrie""",3,0
889,890,"Behr, Mr. Karl Howell",1,1
